interfaces/apparmor/spec.go

```diff
@@ -88,6 +90,92 @@ func RegisteredSnippetKeys() []string {
+// at all will result in a panic, though a tag may be re-registered in
+// association with the same interface with which it was originally registered.
+//
+// Likewise, once registered as unassociated with any interface, attempts to
+// re-register the same tag in association with an interface will result in a
+// panic, but the tag may be re-registered as unassociated again.
+type MetadataTag struct {
+	tag string
+}
+
+func (mt *MetadataTag) String() string {
+	return mt.tag
+}
+
+var validMetadataTagRegexp = regexp.MustCompile("^[a-zA-Z0-9/_-]+$")
+
+func newMetadataTag(tag string) MetadataTag {
+	if !validMetadataTagRegexp.MatchString(tag) {
+		logger.Panicf("cannot register invalid metadata tag: %q", tag)
+	}
+	return MetadataTag{tag: tag}
+}
+
+// registeredMetadataTags is a mapping from metadata tag to the interface with
+// which it was associated when registered. If the tag was registered as
+// unassociated with any interface, its value in the map is "".
+var registeredMetadataTags map[MetadataTag]string = make(map[MetadataTag]string)
+
+const tagUnassociated = ""
+
+// RegisterMetadataTagWithInterface marks the given tag as being associated
+// with the given interface and returns the tag as a MetadataTag. If the tag
+// is already associated with a different interface, or it is explicitly
+// registered as unassociated with any interface, panics with an error.
+func RegisterMetadataTagWithInterface(tag string, iface string) MetadataTag {
+	if iface == "" {
+		logger.Panicf("cannot register metadata tag %q to empty string; use RegisterMetadataTagWithoutInterface instead", tag)
+	}
+	metadataTag := newMetadataTag(tag)
+	if val, ok := registeredMetadataTags[metadataTag]; ok && val != iface {
+		if val == tagUnassociated {
+			logger.Panicf("cannot register metadata tag %q to interface %q when it was previously registered as unassociated", tag, iface)
+		} else {
+			logger.Panicf("cannot register metadata tag %q to two different interfaces: %q and %q", tag, val, iface)
+		}
+	}
+	registeredMetadataTags[metadataTag] = iface
+	return metadataTag
+}
+
+// RegisterMetadataTagWithoutInterface marks the given tag as being unassociated
+// with any interface and returns the tag as a MetadataTag. If the tag is
+// already associated with an interface, panics with an error.
+func RegisterMetadataTagWithoutInterface(tag string) MetadataTag {
+	metadataTag := newMetadataTag(tag)
+	if val, ok := registeredMetadataTags[metadataTag]; ok && val != tagUnassociated {
+		logger.Panicf("cannot register metadata tag %q as unassociated when it was previously associated with interface %q", tag, val)
+	}
+	registeredMetadataTags[metadataTag] = tagUnassociated
+	return metadataTag
+}
+
+// InterfaceForMetadataTag retrieves the interface associated with the given
+// tag, if the tag was registered as associated with an interface. If the tag
+// was not registered, or was registered as unassociated with any interface,
+// returns false.
```

Comment:
How one finds out then if a tag was registered at all but not with an interface?